lib/browser/mqtt.ts

```diff
@@ -390,13 +456,51 @@ export class MqttClientConnection extends BufferedEventEmitter {
         setTimeout(() => { this.uncork() }, 0);
         return new Promise<boolean>((resolve, reject) => {
             const on_connect_error = (error: Error) => {
-                reject(new CrtError(error));
+                let crtError = new CrtError(error);
+                let failureCallbackData = { error: crtError } as OnConnectionFailedResult;
+                this.emit('connection_failure', failureCallbackData);
+                reject(crtError);
             };
+            this.connection.once('error', on_connect_error);
+
+            /**
+             * Annoyingly, MQTT.JS does NOT return socket errors or connection failures through websockets on the browser! This
+             * is because the 'error' result from the websocket 'stream' is not bubbled up and so we have no idea we failed to connect.
+             * This means we have to catch this error manually and send it up ourselves to catch the error.
+             */
+            const on_stream_connect_error = (error: Error) => {
```

Comment:
Annoying, this is the only way to get whether the connection actually failed with MQTT.js websockets. It is not perfect, but it's better than nothing at all.

References:
* https://github.com/mqttjs/MQTT.js/issues/876
* https://stackoverflow.com/questions/49413553/mqtt-js-how-to-handle-connection-errors
* https://github.com/mqttjs/MQTT.js/issues/1282

Note that this will only catch the first failed connection, subsequent failed re-connection attempts will not be caught.

Reply:
I don't agree.  I just checked the 5 client which only uses error to store the particular error.  It does not emit anything; it only emits on closed events.  The closed events are 1-1 with actual connection failures.

Reply:
I did some digging and you are right! The `closed` event is the one that needs to be followed, NOT the `error` event (despite MQTT.js saying it's called with ConnAck packets). I did some looking into how MQTT5 does it and realized I was looking in the wrong place (`error`) and was using the wrong states when I was trying to do it in the right place (`close`) when experimenting.

In other words: I was expecting `error` to be invoked when there was a problem and not `close`, and then when I tried to mimic MQTT5 locally, I had the state checking wrong...

Anyway - it is fixed now! Now it checks to see if a connection was called and will emit the connection failure accordingly based on `close` and whether the client intended to connect. I did a test with MQTT5 and compared it to this, and now this PR emits like MQTT5.